homeassistant/components/fritz/common.py

```diff
@@ -224,29 +229,95 @@ def scan_devices(self, now: datetime | None = None) -> None:
+        self, service_call: ServiceCall, config_entry: ConfigEntry
+    ) -> None:
         """Define FRITZ!Box services."""
-        _LOGGER.debug("FRITZ!Box router: %s", service)
+        _LOGGER.debug("FRITZ!Box router: %s", service_call.service)
 
         if not self.connection:
             raise HomeAssistantError("Unable to establish a connection")
 
+        mode = service_call.data["mode"]
         try:
-            if service == SERVICE_REBOOT:
+            if service_call.service == SERVICE_REBOOT:
                 await self.hass.async_add_executor_job(
                     self.connection.call_action, "DeviceConfig1", "Reboot"
                 )
-            elif service == SERVICE_RECONNECT:
+                return
+
+            if service_call.service == SERVICE_RECONNECT:
                 await self.hass.async_add_executor_job(
                     self.connection.call_action,
                     "WANIPConn1",
                     "ForceTermination",
                 )
+                return
+
+            if service_call.service == SERVICE_CLEANUP and mode == "stale":
+                device_hosts_list: list = await self.hass.async_add_executor_job(
+                    self.fritz_hosts.get_hosts_info
+                )
+                ha_hosts_list: ValuesView[
+                    FritzDevice
+                ] = await self.hass.async_add_executor_job(
+                    lambda: self.devices.values()  # pylint: disable=unnecessary-lambda
+                )
+
         except (FritzServiceError, FritzActionError) as ex:
             raise HomeAssistantError("Service or parameter unknown") from ex
         except FritzConnectionException as ex:
             raise HomeAssistantError("Service not supported") from ex
 
+        if mode == "stale":
+            dev_found: bool = False
+            dev_obj = None
+            for ha_host in ha_hosts_list:
+                for device in device_hosts_list:
+                    if (
+                        ha_host.hostname == device["name"]
+                        and ha_host.mac_address == device["mac"]
+                    ):
+                        dev_found = True
+                        dev_obj = device
+                        break
+                if not dev_found and dev_obj:
+                    _LOGGER.warning(
+                        "Device %s [%s] deleted",
+                        ha_host.hostname,
+                        ha_host.mac_address,
+                    )
+                    entity_reg = (
+                        await self.hass.helpers.entity_registry.async_get_registry()
+                    )
+                    entity_reg.async_remove(dev_obj.entity_id)
+                else:
+                    _LOGGER.debug(
+                        "Device %s [%s] present on Fritz!Box device, nothing to do",
+                        ha_host.hostname,
+                        ha_host.mac_address,
+                    )
+        elif mode == "disabled":
```

Comment:
maybe add a variable for "disabled"?

Reply:
I though about it, but in the end avoided as you cannot use variable in the `services.yaml` anyway.

Reply:
makes sense.